New upload prefix

kaelantdt opened the pull request.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -606,9 +606,19 @@ static void parseDetails(long postNumber, final String rawPost,
+            index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+            prefixLength= Constants.UPLOAD_URI_PREFIX.length();
+        }
+        else {
+            index = line.indexOf(Constants.WEB_CSV_PREFIX);
+            prefixLength= Constants.WEB_CSV_PREFIX.length();
+        }
         assert index != -1 : line;
```
The original developer was being lazy here and making assumptions that the calling code would always be well behaved.  Could you please update this method,  shortURLUploadResponse(), and downloadFileName() replacing the assertion with an explicit check and throw MemberDataException if we haven't found the pattern we are looking for? Use one or more static final error messages and MessageFormat.format here and in the accompanying tests.  The tests should low level String based and live in a new test class called HBParserTest. Thank you.

kaelantdt replied:
> Could you please update this method, shortURLUploadResponse(), and downloadFileName() replacing the assertion with an explicit check and throw MemberDataException if we haven't found the pattern we are looking for?

Hello David! While doing this, it occurred to me that it may be cleaner to add an assertion function in HBParser to check the prefix, and call that in all three of these functions. That way, if we ever wanted to change our supported file prefixes, we could just edit that function, instead of everywhere we need to check file prefixes. Additionally, of course, if we ever added more functions in HBParser that needed this functionality, we could just add a call to that function. Let me know what you think.

Amosnomor replied:
Hello Kaelan.  Some stuff about assertions first. 

In stable running production code, assertions are typically disabled, for performance reasons.  They serve almost as documentation for assumptions being made by the code.  In the test environment, production code assertions are enabled. These are distinct from the test assert mechanisms in the test code.  Production code assertions are also sometimes enabled when trying to track down a production bug.

Assertions should not be used for dealing with user provided input, which can never be trusted to be well behaved. User provided input must always be explicitly checked and errors handled/reported appropriately.  Assertions are not a good mechanism for this because when triggered then cause an immediate error exit from the program.  In our case, running a server, that would mean the server crashing and not reporting anything back to the user.

All that is a long winded way of saying that rather than adding an assertion function to HBParser, let's talk about it as an auditing method, which would thrown an exception (MemberDataException) for problems.

Need to tend to something here at the moment.  Will get back to answering the gist of your suggestion in about 1/2 hour or so.

Amosnomor replied:
Having a single place that encapsulates looking for an uploaded file URL is a good idea.  The UploadFile class is the container for this data, but parsing and creating it is spread out a bit due to organic growth within the code without cleanup pass in this area.  I think that we could hid this logic in a couple of static utility methods within UploadFile.  Something like:
```
   public static boolean containsUploadedFileURL(String data);

    // throws MemberDataException if it can't find a valid URL
   public static UploadFile createUploadFile(String data);

Take a look through the usages of UploadFile throughout the code and let me know what you think.

Amosnomor commented on src/test/java/org/helpberkeley/memberdata/MainTest.java:
```diff
@@ -570,6 +585,31 @@ private void completedOneKitchenOrders(String filepath) throws IOException {
+                        + "\"rows\": [ "
+                        + "[ 42, null, \""
+                        + yesterdayStr
+                        + "\n[xyzzy.csv|attachment](https://go.helpberkeley.org/uploads/short-url/" + filepath + ") (5.8 KB)\" ] "
```
Use Constants.WEB_CSV_PREFIX.  Avoid duplicating strings in code when possible to reduce brittleness (e.g. tests breaking when the constant is changed)

kaelantdt commented:
Hello David! I updated the uploadFile class to contain the audit method as part of construction and updated everywhere uploadFile was used. The only issue I had was in one of the uses of the uploadFile constructor in the DeliveryData class, called in HBParser.dailyDeliveryPosts. This constructor requires that the uploadFile constructor is called with different fields for fileName and shortURL.

To deal with this edge case, for now I have kept the uploadFile constructor public, though I would prefer to find another solution. It seems that this function HBParser.dailyDeliveryPosts is not actually used anymore. Given that, my proposed solutions would be to make the uploadFile constructor private, and implement one of the following:
 - Make a new method to call the constructor `public static uploadFile createUploadFile(String fileName, String shortURL)` for use only in this function
 - Consider this edge case outdated and archive/delete HBParser.dailyDeliveryPosts
 - Change HBParser.dailyDeliveryPosts to not need to create an UploadFile object with both parameters. I tried to do this initially, but since dailyDeliveryPosts is no longer in use, I did not understand exactly what kind of csv data we are parsing, and how it is tokenized. I would either need guidance, or more time to figure out how it works myself. 

Let me know what you think of my changes as a whole, as well as your thoughts on this issue.

Amosnomor commented:
> Hello David! I updated the uploadFile class to contain the audit method as part of construction and updated everywhere uploadFile was used. The only issue I had was in one of the uses of the uploadFile constructor in the DeliveryData class, called in HBParser.dailyDeliveryPosts...
> ...
> Given that, my proposed solutions would be to make the uploadFile constructor private, and implement one of the following:
>     * Consider this edge case outdated and archive/delete HBParser.dailyDeliveryPosts

Yes.  Go ahead and delete HBParser.dailyDeliveryPosts, both signatures, and then also delete the DeliveryData class.

I have some other comments, but it will wait for your next commit.

Also, I apologize for the lack of commenting in the project. Class and method descriptions in particular.  Something that would've have helped would be a description of the Upload class as being used to upload files to Discourse, and the UploadFile class being a container for describing a file that has been uploaded to Discourse.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -25,11 +25,13 @@
 import com.cedarsoftware.io.ReadOptionsBuilder;
 import com.opencsv.CSVReader;
 import com.opencsv.exceptions.CsvException;
+import org.helpberkeley.memberdata.v300.ControlBlockV300;
```
Remove unused imports.  Use Intellij analyze to find these.  Intellij also displays them greyed out if they are unused.

Also, with the exceptions of create methods (e.g. WorkflowParser.create()) and Main methods where we are steering different version/functionality logic, the memberdata package layer should not be looking at the packages that extend/implement it (v200 and v300 packages).

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -25,11 +25,13 @@
 
 import java.io.IOException;
 import java.io.StringReader;
+import java.text.MessageFormat;
```
Remove

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -606,17 +559,32 @@ static void parseDetails(long postNumber, final String rawPost,
 
     public static String shortURLDiscoursePost(final String line) {
-        int index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+        UploadFile.auditFilePrefix(line);
+        int index = -1;
+        int prefixLength = 0;
+        if (line.contains(Constants.UPLOAD_URI_PREFIX)) {
+            index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+            prefixLength= Constants.UPLOAD_URI_PREFIX.length();
+        }
+        else if (line.contains(Constants.WEB_CSV_PREFIX)){
+            index = line.indexOf(Constants.WEB_CSV_PREFIX);
+            prefixLength= Constants.WEB_CSV_PREFIX.length();
+        }
         assert index != -1 : line;
-        String shortURL = line.substring(index);
+        String shortURL = Constants.UPLOAD_URI_PREFIX.concat(line.substring(index + prefixLength));
         index = shortURL.indexOf(')');
         shortURL = shortURL.substring(0, index);
 
         return shortURL;
     }
```
This is only used by UploadFile.createUploadFile().  Move this to be a private method in UploadFile.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -606,17 +559,32 @@ static void parseDetails(long postNumber, final String rawPost,
     }
 
     static String shortURLUploadResponse(final String line) {
-        int index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+        UploadFile.auditFilePrefix(line);
+        int index = -1;
+        if(line.contains(Constants.UPLOAD_URI_PREFIX))
+            index = line.indexOf(Constants.UPLOAD_URI_PREFIX);
+        else if (line.contains(Constants.WEB_CSV_PREFIX))
+            index = line.indexOf(Constants.WEB_CSV_PREFIX);
         assert index != -1 : line;
         String shortURL = line.substring(index);
         index = shortURL.indexOf('"');
```
This is doing some hand-parsing of a JSON response to an upload done by the software.  This would cleaner if we did our regular JSON parsing on it and created an UploadResponse object.   Then we would use that to construct a new UploadFile object.  How about I add parsing for UploadResponse and then you can add UploadFile.createUploadFile(UploadResponse)?   I think that this would better hide the parsing, and normalize higher level usage of this by always dealing with UploadFile instances in the higher level code.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -697,10 +666,9 @@ static OrderHistoryPost orderHistoryPost(final String rawOrderHistoryPost) {
-        String fileName = downloadFileName(lines[2]);
 
-        return new OrderHistoryPost(date, fileName, shortURL);
+
```
Remove. Don't add extra whitespace unintentionally.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -713,9 +681,7 @@ static RestaurantTemplatePost restaurantTemplatePost(final String rawPost) {
 
         for (String line : rawPost.split("\n")) {
             if (line.contains(Constants.UPLOAD_URI_PREFIX)) {
```
This is a spot where we are also exposed to alternate upload URLs.   Add boolean static on UploadFile that checks for both and use it here.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -725,11 +691,8 @@ static RestaurantTemplatePost restaurantTemplatePost(final String rawPost) {
-                String fileName = downloadFileName(line);
-
-                return new UploadFile(fileName, shortURL);
+            if (line.contains(Constants.UPLOAD_URI_PREFIX) || line.contains(Constants.WEB_CSV_PREFIX)) {
```
Use UploadFile.containsUploadedFileURL();
This is how we hide the details of what this URL looks like.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/HBParser.java:
```diff
@@ -775,9 +738,16 @@ static OrderHistory orderHistory(final String orderHistoryData) {
     static String fileNameFromShortURL(final String shortURL) {
-        assert shortURL.startsWith(Constants.UPLOAD_URI_PREFIX);
-        assert shortURL.length() > Constants.UPLOAD_URI_PREFIX.length() : shortURL;
-        return shortURL.substring(Constants.UPLOAD_URI_PREFIX.length());
+        assert (shortURL.startsWith(Constants.UPLOAD_URI_PREFIX) || shortURL.startsWith(Constants.WEB_CSV_PREFIX));
+        String prefix;
+        if (shortURL.startsWith(Constants.UPLOAD_URI_PREFIX)) {
+            prefix = Constants.UPLOAD_URI_PREFIX;
+        }
+        else {
+            prefix = Constants.WEB_CSV_PREFIX;
+        }
+        assert shortURL.length() > prefix.length() : shortURL;
+        return shortURL.substring(prefix.length());
```
Move this to private in UploadFile.

Amosnomor replied:
This still needs to be done.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/UploadFile.java:
```diff
@@ -22,6 +22,10 @@
  */
 package org.helpberkeley.memberdata;
 
+import org.helpberkeley.memberdata.v300.ControlBlockV300;
```
Avoid using v300 package at memerdata level.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/UploadFile.java:
```diff
@@ -45,6 +49,22 @@ public final String getOriginalFileName() {
+            int prefixStart = line.lastIndexOf("]")+2;
+            int prefixEnd = line.lastIndexOf("/")+1;
+            throw new MemberDataException(
+                    MessageFormat.format(ControlBlockV300.INVALID_FILE_PREFIX, line.substring(prefixStart, prefixEnd)));
```
Make a static public constant in UploadFile for this.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/UploadFile.java:
```diff
@@ -45,6 +49,22 @@ public final String getOriginalFileName() {
+
+    public static boolean auditFilePrefix (final String line) throws MemberDataException {
+        if (!(line.contains(Constants.UPLOAD_URI_PREFIX) || line.contains(Constants.WEB_CSV_PREFIX))) {
+            int prefixStart = line.lastIndexOf("]")+2;
+            int prefixEnd = line.lastIndexOf("/")+1;
```
Remove this indexing.  It can crash if the user provided link is completely garbage and doesn't contain what you are expecting here.  Just used the entire line in the exception message.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/WorkRequestHandler.java:
```diff
@@ -332,11 +332,8 @@ static Reply parse(Reply lastReply, final List<String> lines) {
-                    String fileName = HBParser.downloadFileName(line);
-
-                    UploadFile uploadFile = new UploadFile(fileName, shortURL);
+                } else if (line.contains(Constants.UPLOAD_URI_PREFIX) || line.contains(Constants.WEB_CSV_PREFIX)) {
```
Replace with UploadFile.containtsUploadeFileURL()

Amosnomor replied:
This still needs to be done.

Amosnomor commented on src/main/java/org/helpberkeley/memberdata/v300/ControlBlockV300.java:
```diff
@@ -66,6 +66,7 @@ public class ControlBlockV300 extends ControlBlock {
             + "Leave the value column empty if there are no alternate choices.\n";
     public static final String INVALID_MESSAGE_FORMAT = "\"{0}\" is not valid for "
             + Constants.CONTROL_BLOCK_MESSAGE_FORMAT + "\n";
+    public static final String INVALID_FILE_PREFIX = "\"{0}\" is not a supported file prefix";
```
Move this to UploadFile.

Amosnomor commented on src/test/java/org/helpberkeley/memberdata/HBParserTest.java:
```diff
@@ -0,0 +1,34 @@
+package org.helpberkeley.memberdata;
+
+import org.helpberkeley.memberdata.v300.ControlBlockV300;
+import org.junit.Test;
+
+import java.text.MessageFormat;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.helpberkeley.memberdata.UploadFile.auditFilePrefix;
+
+
+public class HBParserTest {
+
+    @Test
+    public void auditFilePrefixTestTrue1() {
+        auditFilePrefix("[xyzzy.csv|attachment](" + Constants.UPLOAD_URI_PREFIX + "someRandomFileID" + ") (5.8 KB)");
+    }
+    @Test
+    public void auditFilePrefixTestTrue2() {
+        auditFilePrefix("[xyzzy.csv|attachment](" + Constants.WEB_CSV_PREFIX + "someRandomFileID" + ") (5.8 KB)");
+    }
+    @Test
+    public void auditFilePrefixTestFails() {
+        Throwable thrown = catchThrowable(() -> auditFilePrefix("[xyzzy.csv|attachment](" + "pload://" + "someRandomFileID" + ") (5.8 KB)"));
+        assertThat(thrown).isInstanceOf(MemberDataException.class);
+        assertThat(thrown).hasMessageContaining(MessageFormat.format(ControlBlockV300.INVALID_FILE_PREFIX, "pload://"));
+
+    }
+
+
+
+
+}
```
The functionality being tested is moving to UploadFile.  Move these to UploadFileTest

Amosnomor commented:
I've added an UploadResponse object and HBParser.uploadResponse() to create it.  This can be used directly in Upload to get rid of  HBParser.shortURLUploadResponse().

kaelantdt commented:
Hello David!

I am quite sorry for the radio silence. These last few days I have been onboarding to a new project for my machine learning program (AI studio) that has bled my working hours with meetings. I finished the intense onboarding period this afternoon, and am back to my regular working hours. In the future, I will make sure to notify you in advance of occurrences like this that I know about ahead of time, I apologize again for the lack of communication.

Amosnomor reviewed (commented):
This cleaned up nicely.

There are still a couple of production code places directly referencing the uploaded file constants that should be changed.  They are called out in the review comments. One that isn't (because it wasn't part of this file set) is in Options.java.  The short url checking it does should also be updated.
